httpmw/recovery/recovery.go

```diff
@@ -0,0 +1,43 @@
+package recovery
+
+import (
+	"net/http"
+
+	"github.com/Azure/aks-middleware/httpmw/logging"
+	"github.com/gorilla/mux"
+)
+
+type PanicHandlerFunc func(logger logging.Logger, w http.ResponseWriter, r *http.Request, err interface{})
+
+func defaultPanicHandler(logger logging.Logger, w http.ResponseWriter, r *http.Request, err interface{}) {
+	logger.Error("Panic occurred", "error", err)
+	http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+}
+
+func NewPanicHandling(logger logging.Logger, panicHandler PanicHandlerFunc) mux.MiddlewareFunc {
+	if panicHandler == nil {
+		panicHandler = defaultPanicHandler
+	}
+	return func(next http.Handler) http.Handler {
+		return &panicHandlingMiddleware{
+			next:         next,
+			logger:       logger,
+			panicHandler: panicHandler,
+		}
+	}
+}
+
+type panicHandlingMiddleware struct {
+	next         http.Handler
+	logger       logging.Logger
+	panicHandler PanicHandlerFunc
+}
+
+func (p *panicHandlingMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer func() {
```

Comment:
Any difference between defer vs just run the code after the line of p.next.ServeHTTP(w, r)?

Reply:
Without defer, If p.next.ServeHTTP(w, r) or any of the other calls in the chain panics, the execution will stop and the recovery logic will not be effective. 

With the defer, the process will defer to the recovery logic when a panic is encountered, resulting in the panic being handled gracefully.

https://www.jajaldoang.com/post/handle-panic-in-http-server-using-middleware-go/